onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2707,15 +2708,26 @@ static void MatmulWithQuantWeightShapeInference(ONNX_NAMESPACE::InferenceContext
   // TODO: check B shape
 
   const auto& dim_last = a_shape.dim(a_shape.dim_size() - 1);
-  if (dim_last.has_dim_value() && dim_last.dim_value() != K) {
-    fail_shape_inference("Incompatible dimensions for matrix multiplication");
-  }
-
   ONNX_NAMESPACE::TensorShapeProto resultShape;
-  for (int i = 0; i < a_shape.dim_size() - 1; ++i) {
-    *resultShape.add_dim() = a_shape.dim(i);
+  if (transB) {
```

Comment:
nit: you can do something like: 
BVal = transB ? K : N; and then keep rest of the logic same.

Reply:
fixed